host/src/proof.rs

```diff
@@ -0,0 +1,213 @@
+
+use crate::{
+    interfaces::{HostError, HostResult},
+    memory,
+    metrics::{
+        inc_guest_error, inc_guest_success, inc_host_error, observe_guest_time,
+        observe_prepare_input_time, observe_total_time,
+    },
+    server::api::v1::{
+        proof::{get_cached_input, set_cached_input, validate_cache_input},
+        ProofResponse,
+    },
+    Opts, TaskChannelOpts,
+};
+
+pub struct ProofActor {
+    rx: Receiver<TaskChannelOpts>,
+    task_count: usize,
+}
+
+impl ProofActor {
+    pub fn new(rx: Receiver<TaskChannelOpts>, task_count: usize) -> Self {
+        Self { rx, task_count }
+    }
+
+    pub async fn run(&mut self) {
+        let semaphore = Arc::new(Semaphore::new(self.task_count));
+        while let Some(message) = self.rx.recv().await {
+            let permit = Arc::clone(&semaphore).acquire_owned().await;
+            tokio::spawn(async move {
+                let _permit = permit;
+                if let Err(error) = Self::handle_message(message).await {
+                    error!("Worker failed due to: {error:?}");
+                }
+            });
+        }
+    }
+
+    pub async fn handle_message(
+        (proof_request, opts, chain_specs): TaskChannelOpts,
+    ) -> HostResult<()> {
+        let (chain_id, blockhash) = get_task_data(
+            &proof_request.network,
+            proof_request.block_number,
+            &chain_specs,
+        )
+        .await?;
+        let mut manager = get_task_manager(&opts.clone().into());
+        // If we cannot track progress with the task manager we can still do the work so we only trace
+        // the error
+        if manager
+            .update_task_progress(
+                chain_id,
+                blockhash,
+                proof_request.proof_type,
+                Some(proof_request.prover.to_string()),
+                TaskStatus::WorkInProgress,
+                None,
+            )
+            .await
+            .is_err()
+        {
+            error!("Could not update task to work in progress via task manager");
+        }
+
+        match handle_proof(&proof_request, &opts, &chain_specs).await {
+            Ok(result) => {
+                let proof_string = result.proof.unwrap_or_default();
+                let proof = proof_string.as_bytes();
+                // We don't need to fail here even if we cannot store it with task manager because the
+                // work has already been done
+                if manager
+                    .update_task_progress(
+                        chain_id,
+                        blockhash,
+                        proof_request.proof_type,
+                        Some(proof_request.prover.to_string()),
+                        TaskStatus::Success,
+                        Some(proof),
+                    )
+                    .await
+                    .is_err()
+                {
+                    error!("Could not update task progress to success via task manager");
```

Comment:
Should we save the error into taskdb to avoid client's repeating query?? I think if error happens here, the task will have a fixed WorkInProgress status.

Reply:
But the problem here is that we cannot access `task_db` so how could we update the status if there is no access to it?
Of course this was just my initial implementation so we could choose to fail the whole task if we cannot obtain the `task_db`